docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,566 @@
+* Extend support for more signature schemes (e.g. secp256r1, ed25519, ed448, sr25519, etc.)
+* More advanced methods ( Post-Quantum Cryptography (PQC) methods
+* Threshold signatures and encryption
+
+Community Engagement Infrastructure:
+
+* Structure the design with tools and documentation interfaces in mind, enabling a seamless future rollout of resources for developer engagement.
+
+## Proposed architecture
+
+### Introduction
+
+In the proposed architecture, each package is decoupled and isolated. Adding new implementations consist of implementing the required interfaces.
+
+```mermaid
+classDiagram
+
+Keyring <|-- Wallet
+
+SecureStorage <|-- Keyring
+SecureItem <|-- SecureStorage
+CryptoProvider <|-- SecureItem
+
+Hasher <|-- CryptoProvider
+Verifier <|-- CryptoProvider
+Signer <|-- CryptoProvider
+Cipher <|-- CryptoProvider
+Generator <|-- CryptoProvider
+```
+
+### Crypto Provider
+
+*Crypto Providers* serve as a middleware responsible for managing the interaction with various instantiated cryptographic packages. It acts as a centralized controller, encapsulating the API of the crypto modules in a single location.
+Through each Crypto provider, users can access functionality such as signing, verification, encryption, and hashing.
+
+By abstracting the underlying cryptographic functionality, *Crypto providers* enable a modular and extensible architecture. It allows users to easily switch between different cryptographic implementations without impacting the rest of the system.
+
+A `CryptoProvider` can be built from different sources, such as a `SecureItem`, a randomness source, a seed, a mnemonic, or a string.
+Since the underlying representation of a `CryptoProvider` is a protobuf message, it can be easily serialized and stored in a safe manner.
+The storing and retrieval is handled by a `SecureStorage` implementation which will be presented in the next section.
+
+```go
+type ProviderMetadata interface {
+  key string
+  value string
+}
+
+type ICryptoProviderMetadata interface {
+  GetTypeUUID() TypeUUID
+  GetName() string
+  GetMetadata() []ProviderMetadata
+}
+
+type ICryptoProviderBuilder interface {
+  ICryptoProviderMetadata
+  
+  FromSecureItem( item SecureItem ) (ICryptoProvider, error)  
+
+  FromRandomness( source IRandomnessSource ) (ICryptoProvider, error)
+  FromSeed( seed []byte ) (ICryptoProvider, error)
+  FromMnemonic( mnemonic string ) (ICryptoProvider error)
+  FromString( url string ) (ICryptoProvider error)
+}
+
+type ICryptoProvider interface {
+  Proto.Message
+  ICryptoProviderMetadata
+  
+  GetKeys() (PubKey, PrivKey, error)
+  GetSigner() (ISigner, error)
+  GetVerifier() (IVerifier, error)
+  GetCipher() (ICipher, error)
+  GetHasher() (IHasher, error)
+}
+```
+
+#### Signing
+
+Interface responsible for signing a message and returning the generated signature.
+
+```go
+type ISigner interface {
+  Sign(Blob) (Signature, error)
```

Comment:
This interface has no space for parameters. Does it mean the precise algorithm is defined as part of the `SecureItem`? That's good but it should be explicitly documented, as it can be limiting when trying to use the same key with different algorithms (which is generally best avoided, but you should make sure it's not necessary for legacy reasons).

Reply:
Correct, the (fixed) params will be stored in the SecureItem struct. Sure, this all will be documented in detail once we have a working demo app.